### DISCUSSION TITLE
java.util.concurrent.Executor instrumentation?

### DISCUSSION THREAD

**Discussion by rkennedy-mode** [open-telemetry/opentelemetry-java-instrumentation#1988] [Q&A] [answered]
I'm using 0.13.0 beta locally with my Dropwizard 2.x service, which has asynchronous JAX-RS resources (my REST endpoints return `java.util.concurrent.CompletableFuture` responses). The CompletableFuture objects are fulfilled via a `java.util.function.Supplier` using an explicit `java.util.concurrent.Executor`:

```java
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

@Path("/greetings/{name}")
@Produces(MediaType.TEXT_PLAIN)
public class GreetingsResource {
    private final Executor threadPool;

    public GreetingsResource(Executor threadPool) {
        this.threadPool = threadPool;
    }

    @GET
    public CompletableFuture<String> getGreeting(@PathParam("name") String name) {
        return CompletableFuture.supplyAsync(() -> String.format("Hello, %s!", name), threadPool);
    }
}
```

The execution of the Supplier doesn't seem to be resulting in a separate Span. I see (via Honeycomb through the [contrib collector](https://github.com/open-telemetry/opentelemetry-collector-contrib)) the Jetty instrumented incoming request and the JAX-RS instrumented call to GreetingsResource.getGreeting() but nothing for the Supplier run via the Executor:

![image](https://user-images.githubusercontent.com/47543687/103696499-4eed8180-4f53-11eb-82a5-7740543a77e1.png)

I didn't see any mention of instrumenting threads, Executors, ExecutorServices, etc in the project README. So I suppose it's not been implemented. It's not mentioned in the [GA requirements](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/docs/ga-requirements.md) document, either. Is this an oversight or was there an intentional decision to leave off instrumenting asynchronous activity like this?

**Comment by trask** [comment, marked as answer]
Hi @rkennedy-mode!

The javaagent should propagate the tracing context across executors / async activities, so if you make an external http/database/etc call from inside of your callback, that should get tracked correlated with the request.

The problem with capturing a span on every executor execution is that can be very noisy / produce a lot of spans.

If you want to capture an additional span that you think is noteworthy, I recommend using the `@WithSpan` annotation: https://github.com/open-telemetry/opentelemetry-java-instrumentation#configure-a-withspan-annotation (and also worth checking out #1945)

**Reply by rkennedy-mode** [reply]
Aha. Thanks, @trask! It does, indeed, work exactly as you mention. I had just figured that child Spans would be created automatically. But it absolutely makes sense that this isn't the behavior of the auto-instrumentation for the reason you've mentioned.
